### PR TITLE
fix(catalog-backend-module-azure): Fixed issue where specifying a branch for discovery did not work 

### PR DESCRIPTION
**Pull request by RagnarHal**
## Hey, I just made a Pull Request!

Was having trouble getting Discovery to work when specifying a specific branch. Apparently, the Azure DevOps search API will only return hits for the default branch unless the name of the branch to be searched is explicitly provided as a filter, even if the branch to be searched has been indexed and is set as searchable. I honestly think this is a bug in the DevOps API, so I'd probably consider this a workaround. The behavior is the same for the newest version of the API (7.1-preview.1).

This PR contains fixes for both for the Catalog Provider and the Static Location configuration techniques. For the Static Location way of discovery, just include a `&version=GB<branchname>` in the URL. This is the same format as the Azure DevOps UI uses for the address bar, so the URL is copy-pasteable from there.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by awanlin** [comment]
Will help review this tomorrow 👍

File: docs/integrations/azure/discovery.md
Comment (awanlin) [inline docs/integrations/azure/discovery.md:170]:
```suggestion
When using a custom pattern, the target is composed of these parts:
```

I think we should just stop counting at this point ;)

File: plugins/catalog-backend-module-azure/src/processors/AzureDevOpsDiscoveryProcessor.test.ts
Comment (awanlin) [inline plugins/catalog-backend-module-azure/src/processors/AzureDevOpsDiscoveryProcessor.test.ts:77]: We should also test branches with a slash for example: `topic/catalog-info`

File: docs/integrations/azure/discovery.md
Comment (awanlin) [inline docs/integrations/azure/discovery.md:1]: Not sure where this should go - probably in this section: https://backstage.io/docs/integrations/azure/discovery#code-search-feature - but I think we should and some details about how to include other branches to the search index, this could be as simple as pointing to the Microsoft docs on the subject: https://learn.microsoft.com/en-us/azure/devops/project/search/functional-code-search?view=azure-devops#search-git-projects-and-repositories

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks for this contribution @RagnarHal, seems reasonable enough to support this. Left a few comments, mostly around documentation but I do think we should be testing branches with `/` as that has been problematic in the past. If that doesn't work then we'll need to document that fact.

File: plugins/catalog-backend-module-azure/src/processors/AzureDevOpsDiscoveryProcessor.test.ts
Comment (RagnarHal) [reply plugins/catalog-backend-module-azure/src/processors/AzureDevOpsDiscoveryProcessor.test.ts]: Thank you for the feedback, that makes a lot of sense.

I've changed the tests to use a slash in the branch name. I don't think it's necessary to test with branches with no slash if we have this. Happy to include those tests though if you disagree.

I also ran some manual tests, both directly against the DevOps API, and with a local instance of Backstage, and I can confirm that using slashes in the branch name works fine. I found that it takes some time (up to an hour in my experience) for a branch that is added to the search index in Azure DevOps to become searchable. The API does not give any indication that the branch is being indexed and just returns no results, which created a bit of confusion for me, so I also added a note to the docs mentioning this.

File: docs/integrations/azure/discovery.md
Comment (RagnarHal) [reply docs/integrations/azure/discovery.md:1]: There appears to already be some documentation on this, it's under https://backstage.io/docs/integrations/azure/discovery/#installation, in a small `note` section. But it's easy to miss.

Happy to move this somewhere else, or perhaps create a dedicated subsection in `#code-search-feature` explaining this?

**Comment by RagnarHal** [comment]
@awanlin thank you for the review! I've made some changes and replied to your comments :)

File: docs/integrations/azure/discovery.md
Comment (awanlin) [reply docs/integrations/azure/discovery.md:1]: Thanks for pointing that out, I clearly missed it. After reading this page a few times I feel like there are several changes that would make it better but they seem outside of the scope for this PR, lets leave it for now.

**Review by awanlin — APPROVED** [review]
This looks pretty great to me, thanks again @RagnarHal 🚀 

Just need a review and approval from the Maintainers now 👍

**Review by freben — APPROVED** [review]

File: plugins/catalog-backend-module-azure/src/processors/AzureDevOpsDiscoveryProcessor.ts
Comment (RagnarHal) [reply plugins/catalog-backend-module-azure/src/processors/AzureDevOpsDiscoveryProcessor.ts:157]: As far as I can tell, this is always true. In Azure DevOps, you can reference different things with the `version` param, such as tags which would then prefix it with `GT`, or specific commits which uses `GC` (I assume "Git Branch", "Git Tag" and "Git Commit"). I haven't tested the Code Search feature with referencing specific commits or tags though, and my assumption is that it wouldn't work since you can only add branches, not tags or commits, to the search index in Azure DevOps.

The reason we have to remove the prefix here is that the Code Search feature just accepts the branch name and won't return any hits if the prefix is included. Since the URL in the address bar in Azure DevOps always includes the prefix, not doing this would mean that directly copy-pasting the location URL would not work, which could be confusing.

So not prefixing `GB` to the branch name in the `azure-discovery` location configuration should actually work as well, but I decided not to explicitly document that to keep it simple and conform to the Azure DevOps way of doing things as much as possible, to avoid adding unnecessary abstractions that might be constraining later.

**Pull request merged**
